File: javascript/factorizar.js


/* Programa para factorizar un número en favtores primos para usar con node.js
   Ejemplo: para factorizar 360 ejecùtelo como

 node factorizar.js 360 */

/* Función que devuelve la factorización de un número n, como una lista de pares [d,e]
   donde d representa a cada uno de los factores primos del nùmero n y e es el exponente correspondiente 
   
   Nota: el código podrìa optimizarse mucho. No está pensado para ser eficiente, sino para que sea fàcil
   de entender.
   */

/* JavaScript code should be executed in "strict mode" */

"use strict";

/* Esta variable la usamos para acumular los textos que queremos devolver en pantalla como resultado */


var output_text="";

/* Con esta constante establecemos si queremos usar <br> o \n como salto de lìnea */

var  break_char;

function get_break_char()
{
  if (typeof window === 'undefined') {
  /* running under node.js */
    return "\n";
} else {
    return "<br";
}
}

/* Esta función factoriza un entero n y devuelve una lista de pares en un array en la forma [d,e] donde
d es un divisor primo de n y e es el exponente correspondiente en su factorizaciòn */

function factorize(n){
let factors=[];
for (let d=2;d<=n;d++)
 {
    if (n%d==0)
    {
      let e=0;  
      do
      {
        e++;
        let m=n/d;
        output_text += n+"="+m+"*"+d+break_char;
        n=m;
      }
      while (n%d==0);
      let new_factor=[d,e];
      factors.push(new_factor);
    };
 };
 return factors;
};

/* Esta funciòn toma la factorizaciòn de un entero calculada por la función anterior y 
 la convierte a un formato agradable para mostrar en la pantalla */

function convert_to_pretty_factorization(factors_list)
{
 if (factors_list.length==0) return "1";
 let text="";
 for (let i=0;i<factors_list.length;i++)
  {
    if (i>0)
      text += "*";
      let factor=factors_list[i];
      text+=factor[0]+"^"+factor[1];
  };
 return text;
}

function pretty_factorization(n)
{
  let factors_list= factorize(n);
  return convert_to_pretty_factorization(factors_list);
}

/* Main Program code */
/* Es posible un manejor más avanzado la linea de comandos usando comander.js 
 https://github.com/tj/commander.js/  pero tratemos de mantener las cosas lo más simples posibles */

 if (process.argv.length == 3) 
 {
  var numero = process.argv[2];
 }
  else
 {
  console.log('Este programa requiere exactamente un argumento: el nùmero a factorizar.') 
  return 1;
 };
 break_char= get_break_char();
 output_text="Usted introdujo:" + numero+break_char  
 if(isNaN(numero)){
	output_text += numero + " no es un número"+break_char;
   }
  else{
    let factorization=pretty_factorization(numero);
    output_text +="La factorización de "+numero+" es "+factorization+break_char; 
   };   
   console.log(output_text);
   return 0;

File: nim/factorizar.nim


# Programa para factorizar un número en favtores primos - versión en nim

# Función que devuelve la factorización de un número n, como una lista de pares [d,e]
# donde d representa a cada uno de los factores primos del nùmero n y e es el exponente correspondiente 
   
# Nota: el código podrìa optimizarse mucho. No está pensado para ser eficiente, 
# sino para que sea fàcilde entender.
 
# Esta función factoriza un entero n y devuelve una secuencia de pares 
# en la forma [d,e] donde
# d es un divisor primo de n y e es el exponente correspondiente en su factorizaciòn */

import strutils

type
 factor_primo = tuple[primo:int, exponente: int] 

var
  factors_seq:seq[factor_primo]
  s: string
  numero:int

proc factorize(n0:int):seq[factor_primo]=
  var
    n,m,d,e: int
    p: factor_primo
    factors_seq=newSeq[factor_primo](0)
  # El parámetro p es inmutable, así que lo copiamos en una variable local.
  # En teoría, declararlo en var debería tener el mismo efecto pero no me funcionó.
  n=n0 
  for d in countup(2,n):
      e=0;  
      while n mod d == 0:
        e=e+1
        m=n div d
        echo(n,"=",m,"*",d)
        n=m
      if (e>0):
        p.primo = d
        p.exponente = e
        factors_seq.add(p)
  return factors_seq  

proc pretty_factorization(factors_seq:seq[factor_primo])=
    var
      already_printed= false
    if len(factors_seq)==0:
      write(stdout,"1")
    else:
     for p in factors_seq:
      if already_printed:
        write(stdout,"*")
      write(stdout,p.primo,"^",p.exponente)
      already_printed= true
    writeLine(stdout,"")

echo("Ingrese un número para factorizar")
s=readline(stdin)
numero=parseInt(s)
echo("Usted ingresó el número ",numero)
factors_seq=factorize(numero)
pretty_factorization(factors_seq)